### PR TITLE
Split controller OSC off from SuperDirt handshake OSC

### LINKED ISSUE #793: ControlAddr should default to 0.0.0.0 if a target isn't localhost
Tidal now uses the control socket to send OSC *out* as well as receive OSC. This is so that superdirt knows where to send replies. However if superdirt is running on a remote host, this means that the default ControlAddr to listen on of localhost (127.0.0.1:6010) stops OSC from being sent.

The default should probably be to listen on all IPs, i.e. `0.0.0.0`, if there is a target other than 127.0.0.1. However people should be able to set a specific IP if they want.

Relatedly, the oHandshake setting on Target is ignored. If it's set to True, maybe OSC should be sent out on an arbitrary IP.

### PR DESCRIPTION
**Pull request by mindofmatthew**
This is an initial step towards disentangling the controller input logic from the `Stream` module (as discussed in #982). Specifically, it removes the SuperDirt handshake process from the controller thread and instead handles all that per-target using the individual SuperDirt sockets. This adds support for more unusual setups (e.g. one Tidal instance running multiple SuperDirt instances at multiple addresses), but should function identically in most cases.

Fixes #793—this uses totally separate addresses for SuperDirt communication and external controllers, so it's not an issue if one is on localhost and the other isn't.

(I've done some tests on my machine, and everything seems to be working the same as in 1.9.4. I started to write some unit tests, but couldn't figure out how to write tests that didn't actually send things over UDP and got nervous that using real UDP communication could result in flaky tests.)

### CODE REVIEW

**Comment by mindofmatthew** [comment]
Pinging @Zalastax, who's expressed interest in OSC listener/stream logic, but happy to hear from anyone else with thoughts as well

File: src/Sound/Tidal/Stream.hs [hunk 1]
```diff
@@ -252,35 +249,51 @@ startStream config oscmap
...
-sendBndl :: Bool -> (Maybe O.Udp) -> Cx -> O.Bundle -> IO ()
-sendBndl isBusMsg (Just listen) cx bndl = O.sendTo listen (O.Packet_Bundle bndl) (N.addrAddress addr)
+sendBndl :: Bool -> Cx -> O.Bundle -> IO ()
+sendBndl isBusMsg cx bndl = O.sendTo (cxUDP cx) (O.Packet_Bundle bndl) (N.addrAddress addr)
   where addr | isBusMsg && isJust (cxBusAddr cx) = fromJust $ cxBusAddr cx
              | otherwise = cxAddr cx
-sendBndl _ Nothing cx bndl = O.sendBundle (cxUDP cx) bndl
 
-resolve :: String -> String -> IO N.AddrInfo
+resolve :: String -> Int -> IO N.AddrInfo
```
Comment (Zalastax) [inline src/Sound/Tidal/Stream.hs:294]: Would it be a good idea to introduce a type synonyme here?
type Port = Int

File: src/Sound/Tidal/Stream.hs [hunk 2]
```diff
@@ -213,28 +212,27 @@ startStream :: Config -> [(Target, [OSC])] -> IO Stream
...
        tidal_status_string >>= verbose config
        verbose config $ "Listening for external controls on " ++ cCtrlAddr config ++ ":" ++ show (cCtrlPort config)
        listen <- openListener config
 
-       cxs <- mapM (\(target, os) -> do remote_addr <- resolve (oAddress target) (show $ oPort target)
-                                        remote_bus_addr <- if isJust $ oBusPort target
-                                                           then Just <$> resolve (oAddress target) (show $ fromJust $ oBusPort target)
-                                                           else return Nothing
+       cxs <- mapM (\(target, os) -> do remote_addr <- resolve (oAddress target) (oPort target)
+                                        remote_bus_addr <- sequence (resolve (oAddress target) <$> (oBusPort target))
```
Comment (Zalastax) [inline src/Sound/Tidal/Stream.hs:223]: Smart! This uses the fact that Maybe is a Monad so sequence is the same thing as what was done here before.

**Review by Zalastax — APPROVED** [review]
Looks good to me!

**Comment by yaxu** [comment]
Sorry for being a bit slow, somehow missed this, but looks great to me!

**Comment by jarmitage** [comment]
@mindofmatthew which of the 4 points you mentioned in https://github.com/tidalcycles/Tidal/issues/982#issuecomment-1620345088 does this address? I'm guessing 1-2? Good stuff!

**Comment by mindofmatthew** [comment]
While trying to verify that this works correctly, I discovered that _Tidal_ doesn't work correctly (see #1030), so I'm going to try and patch that here as well.

@jarmitage Yes, this basically cleans up the Stream module a bit. It doesn't totally separate the listener (which is my eventual plan), but it splits up two listener functions (external controllers vs. getting info from SuperDirt) that are currently combined.

**Comment by yaxu** [comment]
@mindofmatthew are you happy to merge? 

(I'm splitting out stream into a separate tidal-stream package on the cycseq branch.)

**Comment by mindofmatthew** [comment]
@yaxu Sure! I've tested and confirmed that this works. It does have a bug when SuperDirt has different control busses than 0-127 (#1030), but that's also present in 1.9, and I can address it in a separate PR.

**Comment by yaxu** [comment]
Thanks a lot!

**Pull request merged**